packages/app/src/cli/utilities/developer-platform-client.ts

```diff
@@ -295,3 +298,26 @@ export interface DeveloperPlatformClient {
   devSessionDelete: (input: Omit<DevSessionOptions, 'assetsUrl'>) => Promise<DevSessionDeleteMutation>
   getCreateDevStoreLink: (input: string) => Promise<string>
 }
+
+export function createUnauthorizedHandler(client: DeveloperPlatformClient): UnauthorizedHandler {
+  let unauthorizedRetriesUsed = 0
+  return {
+    type: 'token_refresh',
+    handler: async () => {
+      unauthorizedRetriesUsed++
+      if (unauthorizedRetriesUsed > 1) {
+        return {token: undefined}
+      } else if (client.getTokenRefreshInProgress()) {
```

Comment:
I think this is not quite right. We do make simultaneous API calls -- not often, but it can happen, esp. for Partners API.

I'd expect request `A` and `B` to start, then `A` returns with a 401 and we begin the cycle of refreshing. `B` may also return with a 401, and also want to refresh. We would want B to therefore `await` for A's refresh cycle to complete.

In the prototype we did this by storing a nullable promise on the platform client class representing the refresh cycle. If a request needs to refresh and the promise is unset, set it to a promise resolving to the updated token, and then await that promise's conclusion. If a request needs to refresh and the promise is already set, await it. By making it a property of the client class, its essential a singleton value.

Reply:
Should add though that I like the idea of reusing this logic across the two clients. You could tweak so that there is global state like `{[client: DeveloperPlatformClient]: RefreshTokenState}`, where `type RefreshTokenState = undefined | Promise<string>`. Or have the currentlyRefreshingToken promise be part of the DeveloperPlatformClient interface. Maybe thats nicer